### PR TITLE
Allows failing shards without marking as stale

### PR DESCRIPTION
**Pull request by dnhatn**
Currently when failing a shard we also mark it as stale (eg. remove its
allocationId from from the InSync set). However in some cases, we need 
to be able to fail shards but keep them InSync set. This commit adds
such capability. This is a preparatory change to make the primary-replica
resync less lenient.

Relates #24841

### CODE REVIEW

**Review by ywelsch — COMMENTED** [review]
There are many places, where you can set markAsStale to false (because the shard is not active). I wonder if it's nicer to have a separate `removeInSyncId` method on the `RoutingAllocation` class (delegating to IndexMetaDataUpdater) instead of adding this additional parameter to the `shardFailed` method. There are only two places where we would need to call the `removeInSyncId` method (At the moment, it would be in `CancelAllocationCommand` (could go away later), and in `AllocationService.applyFailedShards`). WDYT?

**Comment by dnhatn** [comment]
@ywelsch, I've replaced the markAsStale argument by a separate method. It's nicer than the previous one. Would you please have another look? Thank you.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 1]
```diff
@@ -507,6 +515,7 @@ public String toString() {
             if (failure != null) {
                 components.add("failure [" + ExceptionsHelper.detailedMessage(failure) + "]");
             }
+            components.add("markAsStale [" + markAsStale + "]");
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: it's a bit odd that shard started now also has this component. Maybe a good time to separate ShardEntry into StartedShardEntry and FailedShardEntry?

Comment (ywelsch) [reply core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: while doing that, you can also convert that class to use final fields by not implementing `readFrom` but instead creating a constructor that takes StreamInput as parameter.

**Review by ywelsch — COMMENTED** [review]
I like this better than the previous version. I've left a few more suggestions. I wonder how we can better test that we're marking the correct shards as stale so that when we add failShards calls in the future, we're sure to also correctly call markAsStale.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java [hunk 2]
```diff
@@ -286,8 +274,17 @@ private Updates changes(ShardId shardId) {
     /**
      * Remove allocation id of this shard from the set of in-sync shard copies
      */
-    private void removeAllocationId(ShardRouting shardRouting) {
-        changes(shardRouting.shardId()).removedAllocationIds.add(shardRouting.allocationId().getId());
+    void removeAllocationId(ShardRouting shardRouting) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java]: I would prefer to do this in a follow-up.

**Comment by dnhatn** [comment]
@ywelsch I've split the ShardEntry into started and failed entries; I will make a follow-up for the method `removeStaleIdsWithoutRouting`. Would you please take another look? Thank you.

**Comment by dnhatn** [comment]
@bleskes, I talked to Yannick; he would like you to have on this.

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 3]
```diff
@@ -111,9 +112,9 @@ public void handleResponse(TransportResponse.Empty response) {
                     @Override
                     public void handleException(TransportException exp) {
                         if (isMasterChannelException(exp)) {
-                            waitForNewMasterAndRetry(actionName, observer, shardEntry, listener, changePredicate);
+                            waitForNewMasterAndRetry(actionName, observer, request, listener, changePredicate);
                         } else {
-                            logger.warn((Supplier<?>) () -> new ParameterizedMessage("{} unexpected failure while sending request [{}] to [{}] for shard entry [{}]", shardEntry.shardId, actionName, masterNode, shardEntry), exp);
+                            logger.warn("unexpected failure while sending request [{}] to [{}] for shard entry [{}]", actionName, masterNode, request);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: why not log the exception here?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java [hunk 4]
```diff
@@ -286,8 +274,17 @@ private Updates changes(ShardId shardId) {
...
      */
-    private void removeAllocationId(ShardRouting shardRouting) {
-        changes(shardRouting.shardId()).removedAllocationIds.add(shardRouting.allocationId().getId());
+    void removeAllocationId(ShardRouting shardRouting) {
+        if (shardRouting.active()) {
+            if (shardRouting.primary()) {
+                Updates updates = changes(shardRouting.shardId());
+                if (updates.firstFailedPrimary == null) {
+                    // more than one primary can be failed (because of batching, primary can be failed, replica promoted and then failed...)
+                    updates.firstFailedPrimary = shardRouting;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java]: I think we should leave the "firstFailedPrimary" logic in the shardFailed method for now. I think that this logic can go away later.

**Review by ywelsch — APPROVED** [review]
I've left two more asks. Please also update the PR description to say that this allows shards to be failed without marking them as stale.

**Comment by dnhatn** [comment]
Thanks @ywelsch for your helpful review. I pushed another commit to address your last comments.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java (hunk 4 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java]: Yep, I put it back to the `shardFailed` method.

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 3 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: Done

**Pull request merged**